invoices/modification_interceptor.go

```diff
@@ -0,0 +1,197 @@
+}
+
+// Set atomically sets the callback function. If a callback is already set, an
+// error is returned. The returned function can be used to reset the callback to
+// nil once the client is done.
+func (s *safeCallback) Set(callback HtlcModifyCallback) (func(), error) {
+	if !s.callback.CompareAndSwap(nil, &callback) {
+		return nil, ErrInterceptorClientAlreadyConnected
+	}
+
+	return func() {
+		s.callback.Store(nil)
+	}, nil
+}
+
+// IsConnected returns true if a client is currently connected.
+func (s *safeCallback) IsConnected() bool {
+	return s.callback.Load() != nil
+}
+
+// Exec executes the callback function if it is set. If the callback is not set,
+// an error is returned.
+func (s *safeCallback) Exec(req HtlcModifyRequest) (*HtlcModifyResponse,
+	error) {
+
+	callback := s.callback.Load()
+	if callback == nil {
+		return nil, ErrInterceptorClientDisconnected
+	}
+
+	return (*callback)(req)
+}
+
+// HtlcModificationInterceptor is a service that intercepts HTLCs that aim to
+// settle an invoice, enabling a subscribed client to modify certain aspects of
+// those HTLCs.
+type HtlcModificationInterceptor struct {
+	// callback is the wrapped client callback function that is called when
+	// an invoice is intercepted. This function gives the client the ability
+	// to determine how the invoice should be settled.
+	callback *safeCallback
+
+	// wg is a wait group for active sessions.
+	wg sync.WaitGroup
+
+	// quit is a channel that is closed when the interceptor is stopped.
+	quit chan struct{}
+}
+
+// NewHtlcModificationInterceptor creates a new HtlcModificationInterceptor.
+func NewHtlcModificationInterceptor() *HtlcModificationInterceptor {
+	return &HtlcModificationInterceptor{
+		callback: &safeCallback{},
+	}
+}
+
+// Intercept generates a new intercept session for the given invoice. The call
+// blocks until the client has responded to the request or an error occurs. The
+// response callback is only called if a session was created in the first place,
+// which is only the case if a client is registered.
+func (s *HtlcModificationInterceptor) Intercept(clientRequest HtlcModifyRequest,
+	responseCallback func(HtlcModifyResponse)) error {
+
+	// If there is no client callback set we will not handle the invoice
+	// further.
+	if !s.callback.IsConnected() {
+		log.Debugf("Not intercepting invoice with circuit key %v, no "+
+			"intercept client connected",
+			clientRequest.ExitHtlcCircuitKey)
+
+		return nil
+	}
+
+	// We'll block until the client has responded to the request or an error
+	// occurs.
+	var (
+		responseChan = make(chan *HtlcModifyResponse, 1)
+		errChan      = make(chan error, 1)
+	)
+
+	// The callback function will block at the client's discretion. We will
+	// therefore execute it in a separate goroutine. We put the callback on
+	// the stack by passing it into the goroutine through a parameter. That
+	// means it will be executed in any case, even when the client has
+	// disconnected in the meantime.
```

Comment:
is this comment still valid? i dont think it is being passed in as a param anymore

Reply:
Correct, the second part of the paragraph was outdated. Removed.